Add 7 new difftests and push constants support

LegNeato opened the pull request:
I wanted to increase coverage, so I worked with AI to bang out some tests and infra changes.

New tests:
- bitwise_ops: bit manipulation operations
- trig_ops: trigonometric functions
- control_flow_complex: nested loops and complex control flow
- vector_swizzle: vector component access and swizzling
- memory_barriers: workgroup memory synchronization
- vector_extract_insert: dynamic vector element access
- push_constants: push constants in compute shaders

Infrastructure changes:
- Add WgpuComputeTestPushConstants for push constant support
- Enable PUSH_CONSTANTS feature in wgpu when needed

Firestar99 commented:
I'd love to have these for testing my const folding in https://github.com/Rust-GPU/rust-gpu/pull/317

Firestar99 commented:
CI Ubuntu reports these failures:
    difftests::lang::core::ops::math_ops
    difftests::lang::core::ops::matrix_ops
    difftests::lang::core::ops::vector_ops

On my machine I get these failures:
    difftests::lang::core::ops::math_ops
    difftests::lang::core::ops::matrix_ops
    
Could this be float inaccuracies between vendors?

LegNeato commented:
Yeah, I think it is floats because they all passed locally. I'll take a look and either round or omit

LegNeato commented:
I added a `compat_round!` macro that handled most of the differences, but could not get all of them. So I added epsilon support when comparing. And then because I was debugging with python scripts I added better debug output when dealing with f32s. And because I did that I reworked all the output cases to match.

This is a lot of code and I don't really expect you to review it...AI wrote a lot of it and I reviewed and tweaked. I figure we'll iterate on main as it is super self-contained. I can probably break it up if you want though!

As an aside, I think some of these should be runtests, as I am not sure what value we get diffing with glsl for all of them. BTW, if test times get long we can shard in CI as theoretically we could have 1 job for each test as they are unrelated.

Firestar99 commented on tests/difftests/lib/src/scaffold/compute/vulkano.rs:
```diff
@@ -0,0 +1,250 @@
+    sync::{self, GpuFuture},
+};
+
+pub struct VulkanoBackend {
```
Any particular reason you choose to add a vulkano backend for the workgroup memory test? 

I've used vulkano quite a bit in the past with rust-gpu and even [made an integration example](https://github.com/Firestar99/rust-gpu-vulkano-example), but I'm actually not the biggest fan nowadays. Primarily, their current sync system is deeply flawed and you can get it to quite easily cause validation errors. For a while, I ran a modded variant of their sync with [fixes](https://github.com/vulkano-rs/vulkano/pull/2515) and [performance optimizations](https://github.com/vulkano-rs/vulkano/pull/2539), until this unmaintainable mess completely broke down for me. They are in the process of replacing it with a new task graph based system, which has been in dev for like 1.5 years and still not fully ready to be used. Putting them in a bit of a weird spot currently. And even if they finish it, I'd actually still recommend wgpu over it since I'm not a fan of full task graphs nor do I think they have the manpower to properly maintain it. 

If you want a plain vulkan impl, let me know and I'll write you one in ash.

LegNeato replied:
The windows version of one test is failing and I can't see why. I am trying to see if it is a naga / wgpu bug, so I am adding the vulkano version to compare. I hope to not land the vulkan version.

LegNeato commented:
Ok, the windows issue is definitely a wgpu + naga (dxc) bug.

LegNeato commented:
I've pulled out the repo for the wgpu/naga bug: https://github.com/LegNeato/wgpu-workgroup-memory-bug

Firestar99 commented:
Why do we need swiftshader on windows?

Firestar99 reviewed (changes requested):
What I love about this PR:
* we print any detected changes into the logs, so you can easily see in CI the detected difference without having to boot up that platform
* epsilon for floating point
* the `ComputeShader` trait to make it easy to switch out shaders
* Skip cause we'll likely need it in more places than this

I wonder if we should separate out the difftest improvements from the actual difftests in separate PRs?

The only major concern I have is that there's a lot of copy-paste between the different variants of a single test case. They usually only differ by the `impl ComputeShader` given as a arg. Doesn't have to be this PR, can be a followup, but I'd like to see those almost identical mains collapse. Make all variants of a single difftest a single crate. Some ideas around how that could be done:
* Features. The runner reads all features of the single difftest crate (via `cargo_metadata`) and runs the test with one feature being active at a time. 
* Args, which I almost prefer. Read from some metadata toml file the variants you want and pass them in as args, via `struct Config`. Then switching between shaders is just a match, not some `#[cfg(...)]` monster. Would require every variant to depend on wgpu, naga and ash, but we need to compile all of them anyway.
3. Metadata could be like github runners to allow cartesian product:
```toml
# skip wgsl and cpu variant
[variants]
shader = ["rust", "wgsl"]
runner = ["wgpu", "ash", "cpu"]
```

LegNeato commented on .github/workflows/ci.yaml:
```diff
@@ -29,6 +29,11 @@ jobs:
+        name: Install Vulkan Runtime with SwiftShader (Windows)
+        uses: NcStudios/VulkanCI@v1.2
+        with:
+          sdkVersion: 1.4.309.0
```
Otherwise we hit https://github.com/LegNeato/wgpu-workgroup-memory-bug, which appears to be a bug on WARP.

LegNeato commented on tests/difftests/lib/src/scaffold/compute/ash.rs:
```diff
@@ -0,0 +1,436 @@
+                .api_version(vk::API_VERSION_1_2);
+
+            let instance_create_info =
+                vk::InstanceCreateInfo::default().application_info(&app_info);
```
I think maybe having them as a default on option and letting tests turn it off will be good.

LegNeato commented on tests/difftests/README.md:
```diff
@@ -137,13 +186,66 @@ cargo difftest --nocapture
+
+- Shows which packages produced different outputs
+- Lists output file paths for manual inspection
+- Groups packages by their output values
```
<img width="780" alt="Screenshot 2025-07-09 at 4 44 30 PM" src="https://github.com/user-attachments/assets/e21343d3-9d09-404a-bf19-34dc4123cda4" />

LegNeato commented:
@Firestar99 can you try the latest on your machine to confirm bumping the epsilon makes the tests pass?

I tried to slice this up into more logical commits but it is painful. I'll try to at least clean the history up a bit.

Firestar99 commented:
The larger epsilon did fix that test, everything passes now!

Firestar99 commented on tests/difftests/lib/src/scaffold/compute/ash.rs:
```diff
@@ -0,0 +1,436 @@
+
+    fn run_compute(
+        &self,
+        spirv_bytes: &[u8],
```
I'd be fine deferring this to a followup as well

LegNeato replied:
Think I did what you were expecting here.

Firestar99 commented:
Just FYI, difftest are not clippy checked in CI, fixed in https://github.com/Rust-GPU/rust-gpu/pull/329

LegNeato commented:
Yep, just been offline. Will try to get to it in the next couple days if no one beats me to it

LegNeato commented:
@Firestar99 should be ready now I think

Firestar99 reviewed (approved):
I resolved https://github.com/Rust-GPU/rust-gpu/pull/321#discussion_r2194730037 myself by changing the `TestMetadata` constructors in https://github.com/Rust-GPU/rust-gpu/pull/321/commits/0e026d48dfbb73e204bcf43f5aa0caff897b7262. If you're fine with this change, feel free to merge.

Firestar99 commented:
I have a difftest for the `const_folding` PR that needs these changes
